From e82197626325481a64981ba9c42618681e404839 Mon Sep 17 00:00:00 2001
From: Dhanavandhana Kannan <quic_dhanavan@quicinc.com>
Date: Wed, 13 Nov 2024 14:18:02 +0530
Subject: [PATCH] UPSTREAM:PROTOCOL:HOSTAPD: Fix advertise RNR info of the
 specified group alone in FILS discovery frame

In FILS discovery request frame in 6 GHz standalone mode, RNR IE
contains info of all the vaps from all the groups. But the expectation
is when grouping is enabled, RNR should advertise the info of the vaps
of that particular group alone as each group has separate Tx vap.

Fix by enabling RNR to advertise the info of vaps of the respective
group alone when grouping is enabled in 6 GHz standalone mode.

Fixes: I3b1e1d6f2006(hostapd: Add multi MBSSID group support)

Patch-dependency: r45-001-hostapd-Add-Multi-MBSSID-group-support.patch
Patch-work: None

Signed-off-by: Dhanavandhana Kannan <quic_dhanavan@quicinc.com>
---
 src/ap/ieee802_11.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 1a84ff8..b5b6edc 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -9083,6 +9083,14 @@ static bool hostapd_skip_rnr(size_t i, struct mbssid_ie_profiles *skip_profiles,
 	if (mld_update && skip_profiles && *match_idx &&
 	    ((*match_idx < skip_profiles->start) || (*match_idx >= skip_profiles->end)))
 		return true;
+
+	/* When 6GHz is in STANDALONE MODE with MULTI_MBSSID_GROUPING enabled
+	 * only the group's RNR info should be reported in FILS discovery, so
+	 * don't include the other group's info in RNR
+	 */
+	if (!mld_update && (get_colocation_mode(reporting_hapd) == STANDALONE_6GHZ) &&
+	    (hostapd_mbssid_get_tx_bss(reporting_hapd) != hostapd_mbssid_get_tx_bss(bss)))
+		return true;
 #endif /* CONFIG_IEEE80211BE */
 
 	return false;
-- 
2.34.1

